# Importing library

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Loading the Dataset



In [3]:
df = pd.read_csv('/content/drive/MyDrive/Loan payments data.csv')

# Display top 5 records

In [4]:
df.head()

Loan_ID loan_status  Principal  terms effective_date   due_date  \
0  xqd20166231     PAIDOFF       1000     30       9/8/2016  10/7/2016   
1  xqd20168902     PAIDOFF       1000     30       9/8/2016  10/7/2016   
2  xqd20160003     PAIDOFF       1000     30       9/8/2016  10/7/2016   
3  xqd20160004     PAIDOFF       1000     15       9/8/2016  9/22/2016   
4  xqd20160005     PAIDOFF       1000     30       9/9/2016  10/8/2016   

     paid_off_time  past_due_days  age             education  Gender  \
0  9/14/2016 19:31            NaN   45  High School or Below    male   
1   10/7/2016 9:00            NaN   50              Bechalor  female   
2  9/25/2016 16:58            NaN   33              Bechalor  female   
3  9/22/2016 20:00            NaN   27               college    male   
4  9/23/2016 21:36            NaN   28               college  female   

   Unnamed: 11  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN

# Display all Datatypes

In [5]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Loan_ID         500 non-null    object 
 1   loan_status     500 non-null    object 
 2   Principal       500 non-null    int64  
 3   terms           500 non-null    int64  
 4   effective_date  500 non-null    object 
 5   due_date        500 non-null    object 
 6   paid_off_time   400 non-null    object 
 7   past_due_days   200 non-null    float64
 8   age             500 non-null    int64  
 9   education       500 non-null    object 
 10  Gender          500 non-null    object 
 11  Unnamed: 11     0 non-null      float64
dtypes: float64(2), int64(3), object(7)
memory usage: 47.0+ KB
None


# 5 Points Summary Data

In [6]:
print(df.describe())

         Principal       terms  past_due_days         age  Unnamed: 11
count   500.000000  500.000000      200.00000  500.000000          0.0
mean    943.200000   22.824000       36.01000   31.116000          NaN
std     115.240274    8.000064       29.38088    6.084784          NaN
min     300.000000    7.000000        1.00000   18.000000          NaN
25%    1000.000000   15.000000        3.00000   27.000000          NaN
50%    1000.000000   30.000000       37.00000   30.000000          NaN
75%    1000.000000   30.000000       60.00000   35.000000          NaN
max    1000.000000   30.000000       76.00000   51.000000          NaN


In [16]:
df.drop('Unnamed: 11', axis=1, inplace=True)

# Handle missing values in numeric columns

In [20]:
df.fillna(df.mean(numeric_only=True), inplace=True)

# Drop the columns which is not useful

In [24]:
X = df.drop(['loan_status', 'Loan_ID', 'effective_date', 'due_date', 'paid_off_time', 'education', 'Gender'], axis=1)
y = df['loan_status']

# Model Train

In [25]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [28]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Create Dummies

In [32]:
y_train = pd.get_dummies(y_train)
y_test = pd.get_dummies(y_test)

# Create the Sequential Model

In [33]:
model = Sequential()
model.add(Dense(64,activation='relu',input_dim=X_train.shape[1]))
model.add(Dense(32,activation='relu'))
model.add(Dense(y_train.shape[1],activation='softmax'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


# Model Compilation

In [34]:
model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])

# Train the model

In [35]:
model.fit(X_train,y_train,epochs=10,batch_size=32,validation_data=(X_test,y_test))

Epoch 1/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 2s 24ms/step - accuracy: 0.5003 - loss: 1.0280 - val_accuracy: 0.5300 - val_loss: 0.9373
Epoch 2/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.6731 - loss: 0.8743 - val_accuracy: 0.5700 - val_loss: 0.8278
Epoch 3/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.6778 - loss: 0.7685 - val_accuracy: 0.7100 - val_loss: 0.7120
Epoch 4/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.8026 - loss: 0.6597 - val_accuracy: 0.8700 - val_loss: 0.5854
Epoch 5/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.8938 - loss: 0.5369 - val_accuracy: 0.9400 - val_loss: 0.4652
Epoch 6/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.9403 - loss: 0.4506 - val_accuracy: 0.9500 - val_loss: 0.3659
Epoch 7/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.9486 - loss: 0.3451 - val_accuracy: 0.9600 - val_loss: 0.2871
Epoch 8/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.9530 - loss: 0.2982 - val_accuracy: 0.9600 - val_loss

# ~Training Accuracy: 94 - 96%
# ~Validation Accuracy: 96%
# ~Training Loss: 0.24
# ~Validation Loss: 0.19

The neural network model was trained using the training dataset (X_train,y_train) for 10 epochs with a batch size of 32. The training process also included validation on the test dataset (X_test,y_test) to monitor performance and prevent overfitting.

In [36]:
y_pred = model.predict(X_test)
y_pred = (y_pred > 0.5).astype('int32')

4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step


In [37]:
# Convert one-hot encoded predictions and true labels back to single labels
y_pred_classes = np.argmax(y_pred, axis=1)
y_test_classes = np.argmax(y_test.values, axis=1)



In [38]:
print('Accuracy:',accuracy_score(y_test_classes,y_pred_classes))
print('Classification Report:')
print(classification_report(y_test_classes,y_pred_classes))
print('Confusion Matrix:')
print(confusion_matrix(y_test_classes,y_pred_classes))

Accuracy: 0.96
Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        24
           1       1.00      0.83      0.91        24
           2       0.93      1.00      0.96        52

    accuracy                           0.96       100
   macro avg       0.98      0.94      0.96       100
weighted avg       0.96      0.96      0.96       100

Confusion Matrix:
[[24  0  0]
 [ 0 20  4]
 [ 0  0 52]]


# ~Accuracy: 0.96
# ~Macro Avg F1-score: 0.95
# ~Weighted Avg F1-score: 0.96

After training the neural network model, it was evaluated on the test dataset to assess its performanc. The model achieved an overall accuracy of 96%, indicating sstrong classes.